[no-unused-vars] v4 false positive: in d.ts files

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JS
module.exports = {
	parser: '@typescript-eslint/parser',
	parserOptions: {
		project: './tsconfig.json'
	},
	extends: [
		'eslint:recommended',
		'plugin:@typescript-eslint/recommended',
		'plugin:@typescript-eslint/recommended-requiring-type-checking',
	],
	rules: {
		'@typescript-eslint/explicit-function-return-type': 'off',
		'@typescript-eslint/indent': ['warn', 'tab'],
	}
};
```

**foo.d.ts**:
```TS
interface Foo {
	foo: boolean,
}
```

```JSON
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": false,
    "outDir": "./dist",
    "outFile": "./dist/index.js",
    "moduleResolution": "node",
    "target": "es5",
    "module": "system",
    "sourceMap": true,
    "baseUrl": "./"
  },
  "compileOnSave": false,
  "include": [
    "*"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

**Expected Result**

in a `d.ts` file all interfaces are implicitly exported and therefore 'used'

**Actual Result**

  1:11  warning  'Foo' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)

**Additional Info**

no

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.0` |
| `node`                             | `14.8.0` |

--- a/.cspell.json
+++ b/.cspell.json
@@ -38,6 +38,7 @@
   "words": [
     "Airbnb",
     "Airbnb's",
+    "ambiently",
     "ASTs",
     "autofix",
     "autofixers",

--- a/packages/eslint-plugin/src/rules/no-unused-vars.ts
+++ b/packages/eslint-plugin/src/rules/no-unused-vars.ts
@@ -28,6 +28,7 @@ export default util.createRule<Options, MessageIds>({
   defaultOptions: [{}],
   create(context) {
     const rules = baseRule.create(context);
+    const filename = context.getFilename();
 
     /**
      * Gets a list of TS module definitions for a specified variable.
@@ -207,19 +208,93 @@ export default util.createRule<Options, MessageIds>({
         }
       },
 
-      // TODO - this could probably be refined a bit
-      '*[declare=true] Identifier'(node: TSESTree.Identifier): void {
-        context.markVariableAsUsed(node.name);
-        const scope = context.getScope();
-        const { variableScope } = scope;
-        if (variableScope !== scope) {
-          const superVar = variableScope.set.get(node.name);
+      // declaration file handling
+      [declarationSelector(AST_NODE_TYPES.Program, true)](
+        node: DeclarationSelectorNode,
+      ): void {
+        if (!util.isDefinitionFile(filename)) {
+          return;
+        }
+        markDeclarationChildAsUsed(node);
+      },
+
+      // declared namespace handling
+      [declarationSelector(
+        'TSModuleDeclaration[declare = true] > TSModuleBlock',
+        false,
+      )](node: DeclarationSelectorNode): void {
+        markDeclarationChildAsUsed(node);
+      },
+    };
+
+    type DeclarationSelectorNode =
+      | TSESTree.TSInterfaceDeclaration
+      | TSESTree.TSTypeAliasDeclaration
+      | TSESTree.ClassDeclaration
+      | TSESTree.FunctionDeclaration
+      | TSESTree.TSDeclareFunction
+      | TSESTree.TSEnumDeclaration
+      | TSESTree.TSModuleDeclaration
+      | TSESTree.VariableDeclaration;
+    function declarationSelector(
+      parent: string,
+      childDeclare: boolean,
+    ): string {
+      return [
+        // Types are ambiently exported
+        `${parent} > :matches(${[
+          AST_NODE_TYPES.TSInterfaceDeclaration,
+          AST_NODE_TYPES.TSTypeAliasDeclaration,
+        ].join(', ')})`,
+        // Value things are ambiently exported if they are "declare"d
+        `${parent} > :matches(${[
+          AST_NODE_TYPES.ClassDeclaration,
+          AST_NODE_TYPES.TSDeclareFunction,
+          AST_NODE_TYPES.TSEnumDeclaration,
+          AST_NODE_TYPES.TSModuleDeclaration,
+          AST_NODE_TYPES.VariableDeclaration,
+        ].join(', ')})${childDeclare ? '[declare=true]' : ''}`,
+      ].join(', ');
+    }
+    function markDeclarationChildAsUsed(node: DeclarationSelectorNode): void {
+      const identifiers: TSESTree.Identifier[] = [];
+      switch (node.type) {
+        case AST_NODE_TYPES.TSInterfaceDeclaration:
+        case AST_NODE_TYPES.TSTypeAliasDeclaration:
+        case AST_NODE_TYPES.ClassDeclaration:
+        case AST_NODE_TYPES.FunctionDeclaration:
+        case AST_NODE_TYPES.TSDeclareFunction:
+        case AST_NODE_TYPES.TSEnumDeclaration:
+        case AST_NODE_TYPES.TSModuleDeclaration:
+          if (node.id?.type === AST_NODE_TYPES.Identifier) {
+            identifiers.push(node.id);
+          }
+          break;
+
+        case AST_NODE_TYPES.VariableDeclaration:
+          for (const declaration of node.declarations) {
+            visitPattern(declaration, pattern => {
+              identifiers.push(pattern);
+            });
+          }
+          break;
+      }
+
+      const scope = context.getScope();
+      const { variableScope } = scope;
+      if (variableScope !== scope) {
+        for (const id of identifiers) {
+          const superVar = variableScope.set.get(id.name);
           if (superVar) {
             superVar.eslintUsed = true;
           }
         }
-      },
-    };
+      } else {
+        for (const id of identifiers) {
+          context.markVariableAsUsed(id.name);
+        }
+      }
+    }
 
     function visitPattern(
       node: TSESTree.Node,

--- a/packages/eslint-plugin/tests/rules/no-unused-vars.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unused-vars.test.ts
@@ -551,20 +551,6 @@ declare namespace Foo {
 }
 console.log(Foo);
     `,
-    // https://github.com/typescript-eslint/typescript-eslint/issues/61
-    `
-declare var Foo: {
-  new (value?: any): Object;
-  foo(): string;
-};
-    `,
-    // https://github.com/typescript-eslint/typescript-eslint/issues/106
-    `
-declare class Foo {
-  constructor(value?: any): Object;
-  foo(): string;
-}
-    `,
     `
 import foo from 'foo';
 export interface Bar extends foo.i18n {}
@@ -629,9 +615,6 @@ export default class Foo {
     prop();
   }
 }
-    `,
-    `
-declare function foo(a: number): void;
     `,
     `
 export function foo(): void;
@@ -788,6 +771,24 @@ export const StyledPayment = styled.div<StyledPaymentProps>\`\`;
 import type { foo } from './a';
 export type Bar = typeof foo;
     `,
+    // https://github.com/typescript-eslint/typescript-eslint/issues/2456
+    {
+      code: `
+interface Foo {}
+type Bar = {};
+declare class Clazz {}
+declare function func();
+declare enum Enum {}
+declare namespace Name {}
+declare const v1 = 1;
+declare var v2 = 1;
+declare let v3 = 1;
+declare const { v4 };
+declare const { v4: v5 };
+declare const [v6];
+      `,
+      filename: 'foo.d.ts',
+    },
   ],
 
   invalid: [

